Sources/SwiftSyntax/SyntaxArena.swift

```diff
@@ -0,0 +1,178 @@
+  /// uninitlialized memory range as a `UnsafeMutableBufferPointer<RawSyntax?>`.
+  func allocateRawSyntaxBuffer(count: Int) -> UnsafeMutableBufferPointer<RawSyntax?> {
+    return allocator.allocate(RawSyntax?.self, count: count)
+  }
+
+  /// Allcates a buffer of `RawTriviaPiece` with the given count, then returns
+  /// the uninitialized memory range as a `UnsafeMutableBufferPointer<RawTriviaPiece>`.
+  func allocateRawTriviaPieceBuffer(
+    count: Int) -> UnsafeMutableBufferPointer<RawTriviaPiece> {
+      return allocator.allocate(RawTriviaPiece.self, count: count)
+    }
+
+  /// Allcates a buffer of `UInt8` with the given count, then returns the
+  /// uninitialized memory range as a `UnsafeMutableBufferPointer<UInt8>`.
+  func allocateTextBuffer(count: Int) -> UnsafeMutableBufferPointer<UInt8> {
+    return allocator.allocate(UInt8.self, count: count)
+  }
+
+  /// Copies the contents of a `SyntaxText` to the memory this arena manages,
+  /// and return the `SyntaxText` in the destiation.
+  func intern(_ value: SyntaxText) -> SyntaxText {
+    // Return the passed-in value if it's
+    // * empty,
+    // * a part of "source buffer", or
+    // * in the memory allocated by this arena
+    if (value.isEmpty || sourceBufferContains(value.baseAddress!) ||
+        allocator.contains(address: value.baseAddress!)) {
+      return value
+    }
+
+    let allocated = allocator.allocate(UInt8.self, count: value.count)
+    _ = allocated.initialize(from: value)
+    return SyntaxText(baseAddress: allocated.baseAddress, count: allocated.count)
+  }
+
+  /// Copies a UTF8 sequence of `String` to the memory this arena manages, and
+  /// returns the copied string as a `SyntaxText`
+  func intern(_ value: String) -> SyntaxText {
+    if value.isEmpty { return SyntaxText() }
+    var value = value
+    return value.withUTF8 { utf8 in
+      let allocated = allocator.allocate(UInt8.self, count: utf8.count)
+      _ = allocated.initialize(from: utf8)
+      return SyntaxText(baseAddress: allocated.baseAddress, count: utf8.count)
+    }
+  }
+
+  /// Copies a `RawSyntaxData` to the memory this arena manages, and retuns the
+  /// pointer to the destination.
+  func intern(_ value: RawSyntaxData) -> UnsafePointer<RawSyntaxData> {
+    let allocated = allocator.allocate(RawSyntaxData.self, count: 1).baseAddress!
+    allocated.initialize(to: value)
+    return UnsafePointer(allocated)
+  }
+
+  /// Adds an `SyntaxArena` to this arena as a "child". Do nothing if `arenaRef`
+  /// refers `self`.
+  ///
+  /// When an arena added to another arena, it's owned and is never released
+  /// until the parent arena is deinitialized. This can be used when the syntax
+  /// tree managed by this arena want to hold a subtree owned by other arena.
+  /// See also `RawSyntax.layout()`.
+  func addChild(_ arenaRef: SyntaxArenaRef) {
+    if SyntaxArenaRef(self) == arenaRef { return }
+
+    let other = arenaRef.value
+
+    precondition(
```

Comment:
I don't want to call `!other.contains(arena: self)` in "release" build because it's heavy, hence `assert`. 
`!self.hasParent` is super light operation so we can check it even in "release" build, and it _should_ prevent cyclic reference, but I am not 100% sure, so I added the `assert` just in case.

Reply:
OK, it just made me wonder if I mis-understood the purpose of `hasParent` because of this assert. Just so nobody else stumbles across this, do you think it’s worth adding a comment like
```swift
// `precondition(!self.hasParent)` should be sufficient to make sure we don’t add retain cycles between syntax arenas, but to be doubly sure, check the child arenas as well in debug builds.
```